src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -24,38 +24,54 @@
-//      stmt3                    else
-//    endif                        predicates [clone]
-//    stmt4                        loop [clone]
-//  endloop                          stmt1 [clone]
-//                                   stmt3
-//                                   stmt4 [clone]
-//                                 endloop
-//                               endif
+// The invariant, non-loop-exiting test can now be moved before both loops (to only execute it once) and turned into a
+// loop selector If node to select at runtime which unswitched loop version should be executed.
+// - Loop selector true?  Execute the true-path-loop.
+// - Loop selector false? Execute the false-path-loop.
 //
-// Note: the "else" clause may be empty
-
-//------------------------------policy_unswitching-----------------------------
-// Return TRUE or FALSE if the loop should be unswitched
-// (ie. clone loop with an invariant test that does not exit the loop)
-bool IdealLoopTree::policy_unswitching( PhaseIdealLoop *phase ) const {
+// Note that even though an invariant test that exits the loop could also be optimized with Loop Unswitching, it is more
+// efficient to simply peel the loop which achieves the same result in a simpler manner (also see policy_peeling()).
+//
+// The following graph summarizes the Loop Unswitching optimization:
+//
+//                                                                    [Initialized Assertion Predicates]
+//                                                                                    |
+//    [Predicates]                                                         loop selector If (invariant-test)
+//         |                                                                  /                   \
+//    Original Loop                                                       true?                  false?
+//      stmt1                                                             /                         \
+//      if (invariant-test)           UNSWITCHED              [Cloned Parse Predicates]         [Cloned Parse Predicates]
+//        if-path                     =========>              [Cloned Template                  [Cloned Template
+//      else                                                   Assertion Predicates]             Assertion Predicates]
+//        // could be empty                                         |                                  |
+//        [else-path]                                         True-Path-Loop                    False-Path-Loop
+//      stmt2                                                   cloned stmt1                      cloned stmt1
+//    Endloop                                                   cloned if-path                    [cloned else-path]
+//                                                              cloned stmt2                      cloned stmt2
+//                                                            Endloop                           Endloop
```

Comment:
Could it be the other way around: the if-path is empty and the else-path has code? Or are you just trying to suggest that there could be an empty path, and that this would really simplify one of the cloned loops now?

Reply:
Yes, it would be a simplification. But as suggested in the other comment. It might create more confusion than actually helping.